RLM_learn.thy

```diff
@@ -0,0 +1,189 @@
+  have " (1/n)\<ge> 0" by auto
+  then have "convex_on H  (\<lambda> h. (1/n)*(sum (\<lambda>i. l h (S i)) {0..<n}) )" using convex_on_cmul 3
+    by blast
+  then have "convex_on H  (\<lambda> h.(sum (\<lambda>i. l h (S i)) {0..<n})/n )" by auto
+  then show "convex_on H (train_err_loss S n l)"  unfolding train_err_loss_def by blast
+qed
+
+text\<open>Define a locale for cleaner proofs and definitions\<close>
+locale learning_basics1 =
+
+fixes H :: "'a::{euclidean_space} set"
+  and D :: "('b \<times> 'c) pmf"
+  and n :: "nat"
+  and l :: "('a  \<Rightarrow> ('b \<times> 'c) \<Rightarrow> real)"
+  and k :: "real" 
+assumes nnH: "H \<noteq> {}" 
+  and  convH: "convex H"
+  and l_pos: "\<forall>h\<in>H. \<forall>z\<in>D. l h z \<ge> 0"
+  and convll : "\<forall>z \<in> D. convex_on H (\<lambda> h. l h z)"
+  and n_pos: "n\<ge>1"
+ and k_pos : "k>0"
+begin
+
+text \<open>"Regularized Loss Minimization" 
+Its learning rule RLMe chooses for a given training set S an Hypothesis h from H
+which minimizes the training error + regularisation function. \<close>
+
+definition RLM :: "(nat \<Rightarrow> ('b * 'c)) \<Rightarrow> ('a::euclidean_space \<Rightarrow> real) \<Rightarrow> 'a::euclidean_space set" where
+  "RLM S' r = {h. is_arg_min (train_err_loss S' n l + r) (\<lambda>s. s\<in>H) h}"
+
+definition RLMe :: "(nat \<Rightarrow> ('b * 'c)) \<Rightarrow> ('a::euclidean_space \<Rightarrow> real) \<Rightarrow> 'a::euclidean_space" where 
+  "RLMe S' r = (SOME h. h\<in> RLM S' r)"
+
+definition RLMridge :: "(nat \<Rightarrow> ('b * 'c)) \<Rightarrow> real \<Rightarrow> 'a::euclidean_space" where 
+  "RLMridge S' k' = (SOME h. h\<in> RLM S' (\<lambda> x. k' * (norm x)^2))"
+
+fun ridge_fun :: "(nat \<Rightarrow> ('b * 'c)) \<Rightarrow> real \<Rightarrow> ('a \<Rightarrow> real)" where
+  "ridge_fun S' k' = (train_err_loss S' n l + (\<lambda> w. k' * (norm w)^2))" 
+
+definition ridge_argmin :: "(nat \<Rightarrow> ('b * 'c)) \<Rightarrow> real \<Rightarrow> 'a::euclidean_space set" where
+  "ridge_argmin S' k' = {h. is_arg_min (ridge_fun S' k') (\<lambda>s. s\<in>H) h}"
+
+definition ridge_mine :: "(nat \<Rightarrow> ('b * 'c)) \<Rightarrow> real \<Rightarrow> 'a::euclidean_space" where
+  "ridge_mine S' k' = (SOME h. h \<in> ridge_argmin S' k')"
+
+definition swapped_S :: "(nat \<Rightarrow> ('b * 'c)) \<Rightarrow> nat \<Rightarrow> nat  \<Rightarrow> (nat \<Rightarrow> ('b * 'c))" where
+  "swapped_S S' i m =  S'(i:= S' m, m:= S' i) "  
+
+definition truncated_S :: "(nat \<Rightarrow> ('b * 'c))  \<Rightarrow> nat \<Rightarrow>(nat \<Rightarrow> ('b * 'c))" where
+  "truncated_S S' m =  S'(m := undefined) "
+
+text\<open>S_index is a set where the i-th data point in S is replaced with an arbitrary one\<close>
+definition S_index :: "(nat \<Rightarrow> ('b \<times> 'c)) \<Rightarrow> nat \<Rightarrow> ('b \<times> 'c) \<Rightarrow> (nat \<Rightarrow> ('b * 'c))" where
```

Comment:
Yes it is, not sure at the moment. I think I wasn't sure about some implementations and in the end, it turned out that it could be this easy. Oh wow, I am impressed :D